javascript/ql/src/Security/CWE-116/UnsafeHtmlExpansion.ql

```diff
@@ -0,0 +1,67 @@
+/**
+ * @name Unsafe expansion of shorthand HTML tag
+ * @description Using regular expressions to expand shorthand HTML
+ *              tags may lead to cross-site scripting vulnerabilities.
+ * @kind problem
+ * @problem.severity warning
+ * @precision very-high
+ * @id js/unsafe-html-expansion
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-079
+ *       external/cwe/cwe-116
+ */
+
+import javascript
+
+/**
+ * A regular expression that captures the name and content of a shorthand HTML tag such as `<div id='foo'/>`.
+ */
+class ShorthandTagRecognizer extends RegExpLiteral {
+  ShorthandTagRecognizer() {
+    exists(RegExpSequence seq, RegExpGroup name, RegExpGroup content |
+      // `/.../g`
+      this.isGlobal() and
+      this = seq.getLiteral() and
+      // `/<.../`
+      seq.getChild(0).getConstantValue() = "<" and
+      // `/...\/>/`
+      seq.getLastChild().getPredecessor().getConstantValue() = "/" and
+      seq.getLastChild().getConstantValue() = ">" and
+      // `/...((...)...).../`
+      seq.getAChild() = content and
+      content.getNumber() = 1 and
+      name.getNumber() = 2 and
+      name = content.getChild(0).(RegExpSequence).getChild(0) and
+      // `/...(([a-z]+)...).../` or `/...(([a-z][...]*)...).../`
+      exists(RegExpQuantifier quant | name.getAChild*() = quant |
+        quant instanceof RegExpStar or
+        quant instanceof RegExpPlus
+      ) and
+      // `/...((...)[^>]*).../`
+      exists(RegExpCharacterClass lazy |
+        name.getSuccessor().(RegExpStar).getChild(0) = lazy and
+        lazy.isInverted() and
+        lazy.getAChild().getConstantValue() = ">"
+      )
+    )
+  }
+
+  /**
+   * Gets a data flow node that may refer to this regular expression.
+   */
+  DataFlow::SourceNode ref(DataFlow::TypeTracker t) {
+    t.start() and
+    result = this.flow()
+    or
+    exists(DataFlow::TypeTracker t2 | result = ref(t2).track(t2, t))
+  }
+}
+
+from ShorthandTagRecognizer regexp, StringReplaceCall replace
+where
+  regexp.ref(DataFlow::TypeTracker::end()).flowsTo(replace.getArgument(0)) and
+  replace.getRawReplacement().mayHaveStringValue("<$1></$2>")
+select replace,
+  "This HTML tag expansion may disable earlier sanitizations as $@ may match unintended strings.",
```

Comment:
I understand the qhelp is going to have more details, but I wish we could have a more convincing alert message. The languages comes off as a little too uncertain IMO.

My suggestion would be 
```suggestion
  "This HTML tag expansion invalidates prior sanitization as $@ may match part of an attribute value.",
```

I'm not an expert on mXSS exploits so my wording with "attribute values" might be inaccurate, but let's iterate on this a bit.

Reply:
"attribute" is actually too restrictive in general as the mere expansion of some tags is enough to open an XSS vector. But I think it is correct for the sanitization that we have in mind. 

See these tests: https://github.com/jquery/jquery/pull/4685/files#diff-328a9410239da61336e3662fdbfe9bf6R2947-R2962:

In particular this case:

Original, does not trigger `onerror`:
```html
<noscript/><img src=url404 onerror=xss(11)>
```
Expanded, triggers `onerror`:
```html
<noscript></noscript><img src=url404 onerror=xss(11)>
```